library/dnet.c

```diff
@@ -1013,6 +1484,9 @@ static int dnet_process_cmd_with_backend_raw(struct dnet_backend_io *backend, st
 	gettimeofday(&end, NULL);
 	diff = DIFF(start, end);
 
+	/* If there was any error - send ACK to notify client with error code and destroy transaction */
+	if (err)
+		cmd->flags |= DNET_FLAGS_NEED_ACK;
```

Comment:
No command sends ack itself ever (btw, sending ack means sending small packet without data and without DNET_FLAGS_MORE bit), it is always being sent at the very end of `dnet_process_cmd_raw()` for every commands.
IO commands do not send ack at all, since they send data, and that's already a sign, that everything is ok.
They have to send ack when there are errors though. There must be an ack if any command has failed, since in this case command itself doesn't send data.